### PR TITLE
4750 fix issues 107 and 102

### PR DESCRIPTION
**Pull request by njogz**
# Description

This fixes [issue 102](https://github.com/medic/angular10-migration/issues/102) and [issue 107](https://github.com/medic/angular10-migration/issues/107)

medic/cht-core#[number]

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice! Thanks for fixing this! I just have a couple of suggestions inline.

File: webapp/src/ts/modules/contacts/contacts-content.component.html [hunk 1]
```diff
@@ -74,9 +74,9 @@ <h3 translate>{{card.label}}</h3>
       <div class="card children" *ngFor="let group of selectedContact?.children">
         <div class="action-header cell">
           <div class="col-xs-10">
-            <h3 translate>{{group.type.group_key}}</h3>
+            <h3 translate>{{group?.type?.group_key}}</h3>
           </div>
-          <div *ngIf="group.type.person" mm-auth="can_view_tasks" class="col-xs-2 tasks-title" translate>Tasks</div>
+          <div *ngIf="group.type && group.type.person" mm-auth="can_view_tasks" class="col-xs-2 tasks-title" translate>Tasks</div>
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-content.component.html]: Yes. I have made the change.

File: webapp/src/ts/modules/contacts/contacts-content.component.ts [hunk 2]
```diff
@@ -171,7 +171,7 @@ export class ContactsContentComponent implements OnInit, OnDestroy {
           const parentId = this.selectedContact.doc.parent && this.selectedContact.doc.parent._id;
           return this.router.navigate([`/contacts/${parentId}`]);
         }
-        return this.contactsActions.selectContact(change.id, { silent: true });
+        return this.contactsActions.selectContact(this.route.snapshot.params.id, { silent: true });
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-content.component.ts]: Agreed and I have made the change.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
I'm still seeing an error when I have a contact with an not-existent type that has tasks when I load it: 

image

The error appears to be in the `tasks-for-contact` service, which is not changed by this PR. Even the original issue (https://github.com/medic/angular10-migration/issues/102) displays this error.

**Comment by njogz** [comment]
@dianabarsan this issue is slightly different since the linked issue was caused by an incorrect contact id being set as the selected contact. What do you think of simply just catching the error and logging it similar to how it is in master right now (https://github.com/medic/cht-core/blob/master/webapp/src/js/actions/contacts.js#L79)?

**Comment by dianabarsan** [comment]
> @dianabarsan this issue is slightly different since the linked issue was caused by an incorrect contact id being set as the selected contact. What do you think of simply just catching the error and logging it similar to how it is in master right now (https://github.com/medic/cht-core/blob/master/webapp/src/js/actions/contacts.js#L79)?

@njogz not sure what you mean.
The error code clearly states the line number where the error is happening. 
This line:
https://github.com/medic/cht-core/blob/4750-angular-10-take-2/webapp/src/ts/services/tasks-for-contact.service.ts#L20

And it clearly displays a lack of null-checking the `child.type.person` property.

**Comment by njogz** [comment]
@dianabarsan meant that the trigger to the error was slightly different. I have added optional chaining to deal with the specific error. Please have another look.

File: webapp/src/ts/services/tasks-for-contact.service.ts [hunk 3]
```diff
@@ -15,9 +15,9 @@ export class TasksForContactService {
 
   private getIdsForTasks(model) {
     const contactIds = [];
-    if (!model.type.person && model.children) {
+    if (!model?.type?.person && model?.children) {
       model.children.forEach(child => {
-        if (child.type.person && child.contacts && child.contacts.length) {
+        if (child?.type?.person && child?.contacts && child?.contacts?.length) {
```
Comment (dianabarsan) [inline webapp/src/ts/services/tasks-for-contact.service.ts:20]: I think `if (child?.type?.person && child?.contacts?.length)` is sufficient, is it not?

File: webapp/src/ts/services/tasks-for-contact.service.ts [hunk 4]
```diff
@@ -35,18 +35,19 @@ export class TasksForContactService {
...
+        if (type?.person) {
           return true;
         }
 
         // ... or a leaf place type
         return this.contactTypesService
           .getAll()
           .then(types => {
-            const hasChild = types.some(t => !t.person && t.parents && t.parents.includes(type.id));
+            const hasChild = types.some(t => !t.person && t.parents && t.parents.includes(type?.id));
```
Comment (dianabarsan) [inline webapp/src/ts/services/tasks-for-contact.service.ts:46]: Maybe we can make this a little more efficient and check that `type` is actually defined and return false if it's not, before getting all contact types and such, I'd say that even before checking if the type is a person type.

**Review by dianabarsan — COMMENTED** [review]
Nice work. I have two tiny tiny suggestions inline.

File: webapp/src/ts/services/tasks-for-contact.service.ts [hunk 5]
```diff
@@ -35,18 +39,19 @@ export class TasksForContactService {
...
             return !hasChild;
           });
-      });
+      })
+      .catch(err => console.error('Error finding out if tasks are enabled', err));
```
Comment (dianabarsan) [inline webapp/src/ts/services/tasks-for-contact.service.ts:54]: Not sure if we should be catching these. 
Not catching them should mean that a feedback doc is created automatically, which means that the error will be uploaded and be available for us to see. 
Same goes for the other `catch` that's added in this PR.

Comment (njogz) [reply webapp/src/ts/services/tasks-for-contact.service.ts]: Guess that explains why errors were not being caught. Thanks for the context. I have removed the catch blocks.

File: webapp/src/ts/services/tasks-for-contact.service.ts (hunk 4 above)
Comment (njogz) [reply webapp/src/ts/services/tasks-for-contact.service.ts:50]: Good suggestion. Done.

File: webapp/src/ts/services/tasks-for-contact.service.ts (hunk 3 above)
Comment (njogz) [reply webapp/src/ts/services/tasks-for-contact.service.ts]: It is. I have made the update.

**Review by dianabarsan — APPROVED** [review]
👍🏻

**Pull request merged**